**Датафрейм данных клиники, содержащий информацию о лошадях с коликами.**

**Задание 1. Базовое изучение**
Изучить представленный набор данных на основе описания его столбцов и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

**Задание 2. Работа с выбросами**
В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

**Задание 3. Работа с пропусками**
Рассчитать количество выбросов для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [1]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv('horse_data.csv', names = ['surgery', 'age', 'hospital number', 'rectal temperature', 'pulse', 'respiratory rate', 
                                              'temperature of extremities', 'peripheral pulse', 'mucous membranes',
                                             'capillary refill time', 'pain', 'peristalsis', 'abdominal distension',
                                             'nasogastric tube', 'nasogastric reflux', 'nasogastric reflux PH',
                                             'rectal examination', 'abdomen', 'packed cell volume', 'total protein',
                                             'abdominocentesis appearance', 'abdomcentesis total protein', 'outcome',
                                             'surgical lesion', 'type of lesion', 'type of lesion 2', 'type of lesion 3',
                                              'cp_data'], sep = ',', header=0)
data.head()

surgery  age  hospital number rectal temperature pulse respiratory rate  \
0       2    1           530101              38.50    66               28   
1       1    1           534817               39.2    88               20   
2       2    1           530334              38.30    40               24   
3       1    9          5290409              39.10   164               84   
4       2    1           530255              37.30   104               35   

  temperature of extremities peripheral pulse mucous membranes  \
0                          3                3                ?   
1                          ?                ?                4   
2                          1                1                3   
3                          4                1                6   
4                          ?                ?                6   

  capillary refill time  ... packed cell volume total protein  \
0                     2  ...              45.00          8.40   
1                     1  ...                 50            85   
2                     1  ...              33.00          6.70   
3                     2  ...              48.00          7.20   
4                     2  ...              74.00          7.40   

  abdominocentesis appearance abdomcentesis total protein outcome  \
0                           ?                           ?       2   
1                           2                           2       3   
2                           ?                           ?       1   
3                           3                        5.30       2   
4                           ?                           ?       2   

  surgical lesion type of lesion type of lesion 2 type of lesion 3 cp_data  
0               2          11300                0                0       2  
1               2           2208                0                0       2  
2               2              0                0                0       1  
3               1           2208                0                0       1  
4               2           4300                0                0       2  

[5 rows x 28 columns]

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 28 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   surgery                      300 non-null    object
 1   age                          300 non-null    int64 
 2   hospital number              300 non-null    int64 
 3   rectal temperature           300 non-null    object
 4   pulse                        300 non-null    object
 5   respiratory rate             300 non-null    object
 6   temperature of extremities   300 non-null    object
 7   peripheral pulse             300 non-null    object
 8   mucous membranes             300 non-null    object
 9   capillary refill time        300 non-null    object
 10  pain                         300 non-null    object
 11  peristalsis                  300 non-null    object
 12  abdominal distension         300 non-null    object
 13  nasogastric tube             300 no

**Имеем базу из 300 лошадей, проходивших лечение в клинике.**

**На первый взгляд кажется, что столбцы 26 и 27 (тип хирургического вмешательства 2 и 3) не заполнены у всех. В итоге проверки выяснилиось, что у 6 лошадей было 2 операции, а у одной - 3 за один курс лечения.**

In [4]:
data_multy_lesion = data[ (data['type of lesion 2'] != 0) | (data['type of lesion 3'] != 0) ]
data_multy_lesion

surgery  age  hospital number rectal temperature pulse respiratory rate  \
63        2    1           534293               38.2   130               16   
111       1    1           533736               38.3    40               16   
116       1    1           535208               38.0    44               12   
118       1    1           533983               36.5    78               30   
178       1    1           533968                  ?    60               30   
206       1    1           535137               38.5    96               30   
229       1    9           534597               38.5   120               70   

    temperature of extremities peripheral pulse mucous membranes  \
63                           4                3                4   
111                          3                ?                ?   
116                          1                1                1   
118                          1                ?                1   
178                          3                3                4   
206                          2                3                4   
229                          ?                ?                ?   

    capillary refill time  ... packed cell volume total protein  \
63                      2  ...                 65            82   
111                     1  ...                 38            58   
116                     1  ...                 42            65   
118                     1  ...                 34            75   
178                     2  ...                 45            70   
206                     2  ...                 50            65   
229                     ?  ...                 35            54   

    abdominocentesis appearance abdomcentesis total protein outcome  \
63                            2                           2       3   
111                           1                           2       1   
116                           ?                           ?       1   
118                           2                           1       1   
178                           3                           2       1   
206                           ?                           ?       1   
229                           1                           1       1   

    surgical lesion type of lesion type of lesion 2 type of lesion 3 cp_data  
63                2           2209             1400                0       2  
111               1           3111             7111                0       2  
116               1           3205             3111                0       2  
118               1           3112             6112                0       2  
178               1           6111             3111                0       2  
206               1           6111             3112                0       2  
229               1           4205             3111             2209       2  

[7 rows x 28 columns]

**При этом больничный номер в датафрейм может повторяться, если лошадь проходила лечение в больнице несколько раз. Уникальных идентификаторов всего 284. 16 лошадей появлялись в больнице более одного раза.**

In [5]:
len(data['hospital number'].unique())

284

In [6]:
# drop_list - список идентификаторов тех лошадей, что лечились в клинике более одного раза

list_ = []
drop_list = []
for element in data['hospital number']:
    if element in list_:
        drop_list.append(element)
    list_.append(element)


    
print(drop_list)

[529461, 530693, 532349, 529424, 528996, 527916, 5279822, 528151, 528931, 530526, 528729, 528469, 528904, 529796, 527544, 528890]


In [7]:
data.loc[(data['hospital number'].isin (drop_list), 'number of treat')] = 'few'

In [8]:
data[ (data['number of treat'] == 'few')].sort_values('hospital number', ascending=True)

surgery  age  hospital number rectal temperature pulse respiratory rate  \
282       1    1           527544                  ?    70               16   
110       1    1           527544                  ?    70               16   
166       1    1           527916              38.30    40               18   
160       1    1           527916                  ?     ?                ?   
179       2    1           528151              38.50   100                ?   
205       2    1           528151              38.50   100                ?   
235       1    1           528469              38.10    52               24   
73        1    1           528469              38.10    52               24   
94        2    1           528729              37.50    48               30   
234       2    1           528729              37.50    48               30   
291       1    1           528890              38.90    80               44   
48        1    1           528890              38.90    80               44   
59        2    1           528904                  ?    96                ?   
245       2    1           528904                  ?    96                ?   
67        2    1           528931              38.40    54               24   
225       2    1           528931              38.40    54               24   
156       2    1           528996              38.60    52               20   
38        2    1           528996              38.60    52                ?   
71        1    1           529424              37.80    48               28   
143       1    1           529424              37.80    48               28   
91        2    1           529461              40.30   114               36   
54        2    1           529461              40.30   114               36   
272       1    1           529796                  ?   100               30   
34        1    1           529796                  ?   100               30   
233       1    1           530526              38.30    72               30   
26        1    1           530526              38.30    72               30   
78        1    1           530693                  ?     ?                ?   
136       1    1           530693                  ?     ?                ?   
138       2    1           532349                 38    56                ?   
114       1    1           532349               38.4    44               24   
83        1    1          5279822              38.00     ?               24   
182       1    1          5279822              38.00     ?               24   

    temperature of extremities peripheral pulse mucous membranes  \
282                          3                4                5   
110                          3                4                5   
166                          1                1                1   
160                          ?                ?                ?   
179                          3                3                5   
205                          3                3                5   
235                          1                1                5   
73                           1                1                5   
94                           4                1                3   
234                          4                1                3   
291                          3                3                3   
48                           3                3                3   
59                           3                3                3   
245                          3                3                3   
67                           1                1                1   
225                          1                1                1   
156                          1                1                1   
38                           1                1                1   
71                           1                1                1   
143                          1      

**Что можно сказать:**
- все лошади, лечившиеся несколько раз, старше 6 месяцев
- повторное хирургическое вмешательство в ту же область было у 6 лошадей из 16 (смотрим первую цифру в 'type of lesion'). У остальных органы отличались либо не было операций вовсе.
- основные физиологические показатели у одной и той же лошади совпадают, поэтому при подсчете стат.показателей повторяющиеся идентификаторы можно удалить. Сразу же видно выброс, например столбец 'total protein', где показатели не достигают 10, у нескольких пациентов этот показатель выше 70 (мое предположение - забыли проставить точку).

In [9]:
cut_data = data[['surgery', 'age', 'hospital number', 'rectal temperature', 'pulse', 'temperature of extremities',
                 'abdominal distension','total protein']]
cut_data.head()

surgery  age  hospital number rectal temperature pulse  \
0       2    1           530101              38.50    66   
1       1    1           534817               39.2    88   
2       2    1           530334              38.30    40   
3       1    9          5290409              39.10   164   
4       2    1           530255              37.30   104   

  temperature of extremities abdominal distension total protein  
0                          3                    4          8.40  
1                          ?                    2            85  
2                          1                    1          6.70  
3                          4                    4          7.20  
4                          ?                    ?          7.40

**Выбрано 8 параметров. rectal temperature, pulse, total protein - непрерывные величины, все остальные категорийные.**

**Для того, чтобы воспользоваться методом describe, нужно поменять тип данных и сразу решить, а что делать с пропусками, которые в нашем случае выглядят как "?". Поэтому я заменю все вопросительные знаки на Nan, поменяю тип данных на float (при наличии Nan и попытке заменить на int выдается ошибка).**

In [10]:
cut_data = cut_data.replace('?', 'Nan')

In [11]:
cut_data['rectal temperature'] = cut_data['rectal temperature'].astype('float')
cut_data['pulse'] = cut_data['rectal temperature'].astype('float')
cut_data['temperature of extremities'] = cut_data['temperature of extremities'].astype('float')
cut_data['abdominal distension'] = cut_data['abdominal distension'].astype('float')
cut_data['total protein'] = cut_data['total protein'].astype('float')

In [12]:
cut_data.describe()

age  hospital number  rectal temperature       pulse  \
count  300.000000     3.000000e+02          240.000000  240.000000   
mean     1.640000     1.085889e+06           38.167917   38.167917   
std      2.173972     1.529801e+06            0.732289    0.732289   
min      1.000000     5.184760e+05           35.400000   35.400000   
25%      1.000000     5.289040e+05           37.800000   37.800000   
50%      1.000000     5.303055e+05           38.200000   38.200000   
75%      1.000000     5.347275e+05           38.500000   38.500000   
max      9.000000     5.305629e+06           40.800000   40.800000   

       temperature of extremities  abdominal distension  total protein  
count                  244.000000            244.000000     267.000000  
mean                     2.348361              2.266393      24.456929  
std                      1.045054              1.065131      27.475009  
min                      1.000000              1.000000       3.300000  
25%                      1.000000              1.000000       6.500000  
50%                      3.000000              2.000000       7.500000  
75%                      3.000000              3.000000      57.000000  
max                      4.000000              4.000000      89.000000

In [13]:
print(cut_data['rectal temperature'].mode())
print(cut_data['pulse'].mode())
print(cut_data['temperature of extremities'].mode())
print(cut_data['abdominal distension'].mode())
print(cut_data['total protein'].mode())

0    38.0
dtype: float64
0    38.0
dtype: float64
0    3.0
dtype: float64
0    1.0
dtype: float64
0    6.5
1    7.0
dtype: float64


**По таблице выше я могу судить, что выбросы есть только в первом и последнем столбцах (достаточно посмотреть на СКО). Возраст - категорийный показатель и содержит числа 1 и 2, а мелькает и 9 (скорее набрали либо кол-во месяцев либо лет, в любом случае эти цифры относятся к категории 1, поэтому сразу можно заменить значение и ликвидировать этот выброс).**

**Если я разделю данные на две части по показателю 'surgery', то можно будет сделать первые выводы о том, есть ли разница в биологических показателях лошадей, которым необходимо хирургическое вмешательство, а которых лечат и без него.**

In [14]:
cut_data['age'] = cut_data['age'].replace(9, 1)

In [15]:
horses_with_surgery = cut_data[ (cut_data['surgery'] == '1' ) ]
horses_without_surgery = cut_data[ (cut_data['surgery'] == '2' ) ]

In [16]:
horses_with_surgery.describe()

age  hospital number  rectal temperature       pulse  \
count  180.0     1.800000e+02          140.000000  140.000000   
mean     1.0     1.244480e+06           38.143571   38.143571   
std      0.0     1.702911e+06            0.716025    0.716025   
min      1.0     5.216810e+05           35.400000   35.400000   
25%      1.0     5.290248e+05           37.800000   37.800000   
50%      1.0     5.305260e+05           38.100000   38.100000   
75%      1.0     5.349345e+05           38.500000   38.500000   
max      1.0     5.301219e+06           40.800000   40.800000   

       temperature of extremities  abdominal distension  total protein  
count                  140.000000            142.000000     158.000000  
mean                     2.450000              2.492958      25.286709  
std                      1.075061              1.056869      27.887392  
min                      1.000000              1.000000       3.300000  
25%                      1.000000              2.000000       6.525000  
50%                      3.000000              3.000000       7.500000  
75%                      3.000000              3.000000      58.000000  
max                      4.000000              4.000000      89.000000

In [17]:
horses_without_surgery.describe()

age  hospital number  rectal temperature     pulse  \
count  119.0     1.190000e+02            99.00000  99.00000   
mean     1.0     8.506365e+05            38.20404  38.20404   
std      0.0     1.197621e+06             0.76036   0.76036   
min      1.0     5.184760e+05            36.00000  36.00000   
25%      1.0     5.288580e+05            37.80000  37.80000   
50%      1.0     5.300010e+05            38.20000  38.20000   
75%      1.0     5.340940e+05            38.60000  38.60000   
max      1.0     5.305629e+06            40.30000  40.30000   

       temperature of extremities  abdominal distension  total protein  
count                  103.000000            101.000000     108.000000  
mean                     2.203883              1.960396      22.793519  
std                      0.993698              0.999208      26.639347  
min                      1.000000              1.000000       4.700000  
25%                      1.000000              1.000000       6.500000  
50%                      2.000000              2.000000       7.250000  
75%                      3.000000              3.000000      54.500000  
max                      4.000000              4.000000      86.000000

**Кол-во лошадей с хирургическим вмешательством больше. Разницы в средних показателях ректальной температуры, пульса почти нет (а так же в минимальных и максимальныхб да и в распределении процентилей). А вот уровень белка у лошадей с операциями выше. И показатель состояния брюшной полости (abdominal distension - вздутие живота) у лошадей с операциями ближе к трем нежели у второй категории (несмотря на то, что это категорийный показатель, некие выводы можно сделать), по описанию известно, что этот показатель очень важен, и лошадей со значительным вздутием чаще направляют на операцию, чтобы снять давление с брюшной полости.**

**Также хочется сравнить показатель температуры конечностей лошадей с повышенной ректальной температурой и не превышающей 38,5, так как в описании к датафрейм сказано, что эти показатели коррелируются.**

In [18]:
horses_low_temp = cut_data[ (cut_data['rectal temperature'] <= 38.5 ) ]
horses_high_temp = cut_data[ (cut_data['rectal temperature'] > 38.5 ) ]

In [19]:
horses_low_temp.describe()

age  hospital number  rectal temperature       pulse  \
count  182.0     1.820000e+02          182.000000  182.000000   
mean     1.0     1.001299e+06           37.876923   37.876923   
std      0.0     1.423337e+06            0.528749    0.528749   
min      1.0     5.213990e+05           35.400000   35.400000   
25%      1.0     5.287422e+05           37.600000   37.600000   
50%      1.0     5.303630e+05           38.000000   38.000000   
75%      1.0     5.346172e+05           38.300000   38.300000   
max      1.0     5.299629e+06           38.500000   38.500000   

       temperature of extremities  abdominal distension  total protein  
count                  149.000000            153.000000     165.000000  
mean                     2.167785              2.111111      26.733939  
std                      1.061632              1.035905      28.597110  
min                      1.000000              1.000000       4.000000  
25%                      1.000000              1.000000       6.600000  
50%                      2.000000              2.000000       7.500000  
75%                      3.000000              3.000000      60.000000  
max                      4.000000              4.000000      89.000000

In [20]:
horses_high_temp.describe()

age  hospital number  rectal temperature      pulse  \
count  58.0     5.800000e+01           58.000000  58.000000   
mean    1.0     1.515877e+06           39.081034  39.081034   
std     0.0     1.945188e+06            0.491848   0.491848   
min     1.0     5.216810e+05           38.600000  38.600000   
25%     1.0     5.294145e+05           38.700000  38.700000   
50%     1.0     5.305840e+05           38.950000  38.950000   
75%     1.0     5.351728e+05           39.300000  39.300000   
max     1.0     5.305629e+06           40.800000  40.800000   

       temperature of extremities  abdominal distension  total protein  
count                   49.000000             47.000000      54.000000  
mean                     2.469388              2.255319      20.959259  
std                      1.022702              0.988369      25.149437  
min                      1.000000              1.000000       3.300000  
25%                      2.000000              1.500000       6.500000  
50%                      3.000000              2.000000       7.450000  
75%                      3.000000              3.000000      12.025000  
max                      4.000000              4.000000      85.000000

**Зависимость есть, но не та, что ожидалась. Показатель температуры конечностей ближе к трем у лошадей с высокой ректальной температурой, в то время, как 3 - cool, то есть температура конечностей пониженная. Может быть это связано с шоковым состоянием лошади. Что еще можно сказать, при повышении температуры увеличивается пульс и снижается уровень белка.**

**Отмечу, что возраст всех лошадей, находящихся в клинике, больше 6 месяцев.**

**Избавлюсь от выбросов в столбце с уровнем белка. Причем я уверена, что выбросы лишь в максимальных значениях. Выше можно посмотреть моду по протеину - 6.5 и 7.**

In [21]:
print(cut_data['total protein'].mean())
print(cut_data[ cut_data['total protein'] < 80]['total protein'].mean())

24.456928838951317
22.648648648648653


In [22]:
print(cut_data['total protein'].median())
print(cut_data[ cut_data['total protein'] < 80]['total protein'].median())

7.5
7.5


In [23]:
print(cut_data['total protein'].mode())
print(cut_data[ cut_data['total protein'] < 80]['total protein'].mode())

0    6.5
1    7.0
dtype: float64
0    6.5
1    7.0
dtype: float64


**Изменились лишь показатели среднего значения.** 
**Сделаю то же самое через квартили.**

In [24]:
def outliers(column):
    q1 = cut_data[column].quantile(0.25)
    q3 = cut_data[column].quantile(0.75)
    iqr = q3 - q1
    lower_bound = q1 - (1.5 * iqr) 
    upper_bound = q3 + (1.5 * iqr)
    remove_outliers = cut_data[cut_data[column].between(lower_bound, upper_bound, inclusive=True)].sort_values(column)
    return remove_outliers

In [25]:
outliers('total protein')

surgery  age  hospital number  rectal temperature  pulse  \
255       1    1          5294539                38.8   38.8   
210       1    1           530401                37.9   37.9   
84        1    1           529849                37.8   37.8   
142       1    1           530402                 NaN    NaN   
247       1    1           530170                38.1   38.1   
..      ...  ...              ...                 ...    ...   
144       1    1           534519                37.1   37.1   
63        2    1           534293                38.2   38.2   
1         1    1           534817                39.2   39.2   
90        2    1           534197                38.0   38.0   
208       1    1           535338                37.8   37.8   

     temperature of extremities  abdominal distension  total protein  
255                         1.0                   3.0            3.3  
210                         NaN                   2.0            4.0  
84                          1.0                   3.0            4.5  
142                         4.0                   4.0            4.5  
247                         3.0                   3.0            4.6  
..                          ...                   ...            ...  
144                         3.0                   4.0           81.0  
63                          4.0                   4.0           82.0  
1                           NaN                   2.0           85.0  
90                          NaN                   NaN           86.0  
208                         3.0                   3.0           89.0  

[267 rows x 8 columns]

In [26]:
cut_data.sort_values('total protein', ascending=False)

surgery  age  hospital number  rectal temperature  pulse  \
208       1    1           535338                37.8   37.8   
90        2    1           534197                38.0   38.0   
1         1    1           534817                39.2   39.2   
63        2    1           534293                38.2   38.2   
14        1    1           534917                38.2   38.2   
..      ...  ...              ...                 ...    ...   
245       2    1           528904                 NaN    NaN   
248       1    1           527709                38.0   38.0   
274       1    1           534624                 NaN    NaN   
288       1    1           529428                 NaN    NaN   
296       2    1           527702                37.2   37.2   

     temperature of extremities  abdominal distension  total protein  
208                         3.0                   3.0           89.0  
90                          NaN                   NaN           86.0  
1                           NaN                   2.0           85.0  
63                          4.0                   4.0           82.0  
14                          3.0                   1.0           81.0  
..                          ...                   ...            ...  
245                         3.0                   4.0            NaN  
248                         2.0                   3.0            NaN  
274                         NaN                   4.0            NaN  
288                         NaN                   NaN            NaN  
296                         3.0                   3.0            NaN  

[300 rows x 8 columns]

In [27]:
cut_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery                     300 non-null    object 
 1   age                         300 non-null    int64  
 2   hospital number             300 non-null    int64  
 3   rectal temperature          240 non-null    float64
 4   pulse                       240 non-null    float64
 5   temperature of extremities  244 non-null    float64
 6   abdominal distension        244 non-null    float64
 7   total protein               267 non-null    float64
dtypes: float64(5), int64(2), object(1)
memory usage: 18.9+ KB


**Судя по квартилям выбросов в значении белка нет. Показано 267 ненулевых строк (можно свериться с инфо). Нормальный уровень белка в пределах 6-7, поэтому два предположения, либо большое кол-во данных записано с ошибкой либо 38 лошадей с показателем белка, превышающим в 10 раз норму.**

In [28]:
print(cut_data[ cut_data['total protein'] > 65]['total protein'].count())

38


**То же самое сделаю с остальными столбцами.**

**Ректальная температура**

In [29]:
outliers('rectal temperature')

surgery  age  hospital number  rectal temperature  pulse  \
113       1    1           527933                36.8   36.8   
277       2    1           528620                36.9   36.9   
70        2    1           528214                37.0   37.0   
292       1    1           530034                37.0   37.0   
190       1    1           528742                37.1   37.1   
..      ...  ...              ...                 ...    ...   
165       1    1           534925                39.4   39.4   
231       2    1           530028                39.5   39.5   
287       2    1          5305129                39.5   39.5   
117       2    1          5290482                39.5   39.5   
162       2    1           535176                39.5   39.5   

     temperature of extremities  abdominal distension  total protein  
113                         NaN                   NaN            NaN  
277                         2.0                   2.0            6.5  
70                          3.0                   3.0            7.6  
292                         1.0                   3.0            6.9  
190                         NaN                   1.0            6.7  
..                          ...                   ...            ...  
165                         NaN                   3.0           64.0  
231                         3.0                   1.0            6.4  
287                         NaN                   NaN            5.0  
117                         3.0                   3.0            6.7  
162                         3.0                   2.0           56.0  

[226 rows x 8 columns]

In [30]:
cut_data.sort_values('rectal temperature', ascending=False)

surgery  age  hospital number  rectal temperature  pulse  \
259       1    1           527734                40.8   40.8   
54        2    1           529461                40.3   40.3   
91        2    1           529461                40.3   40.3   
281       2    1           530624                40.0   40.0   
20        1    1           530157                39.9   39.9   
..      ...  ...              ...                 ...    ...   
274       1    1           534624                 NaN    NaN   
282       1    1           527544                 NaN    NaN   
288       1    1           529428                 NaN    NaN   
293       1    1           534004                 NaN    NaN   
295       1    1           533886                 NaN    NaN   

     temperature of extremities  abdominal distension  total protein  
259                         3.0                   1.0            7.4  
54                          3.0                   3.0            8.1  
91                          3.0                   3.0            8.1  
281                         3.0                   1.0            6.5  
20                          1.0                   4.0            6.1  
..                          ...                   ...            ...  
274                         NaN                   4.0            NaN  
282                         3.0                   2.0            7.5  
288                         NaN                   NaN            NaN  
293                         3.0                   NaN           62.0  
295                         4.0                   NaN           65.0  

[300 rows x 8 columns]

**Пульс**

In [31]:
outliers('pulse')

surgery  age  hospital number  rectal temperature  pulse  \
113       1    1           527933                36.8   36.8   
277       2    1           528620                36.9   36.9   
70        2    1           528214                37.0   37.0   
292       1    1           530034                37.0   37.0   
190       1    1           528742                37.1   37.1   
..      ...  ...              ...                 ...    ...   
165       1    1           534925                39.4   39.4   
231       2    1           530028                39.5   39.5   
287       2    1          5305129                39.5   39.5   
117       2    1          5290482                39.5   39.5   
162       2    1           535176                39.5   39.5   

     temperature of extremities  abdominal distension  total protein  
113                         NaN                   NaN            NaN  
277                         2.0                   2.0            6.5  
70                          3.0                   3.0            7.6  
292                         1.0                   3.0            6.9  
190                         NaN                   1.0            6.7  
..                          ...                   ...            ...  
165                         NaN                   3.0           64.0  
231                         3.0                   1.0            6.4  
287                         NaN                   NaN            5.0  
117                         3.0                   3.0            6.7  
162                         3.0                   2.0           56.0  

[226 rows x 8 columns]

In [32]:
cut_data.sort_values('pulse', ascending=False)

surgery  age  hospital number  rectal temperature  pulse  \
259       1    1           527734                40.8   40.8   
54        2    1           529461                40.3   40.3   
91        2    1           529461                40.3   40.3   
281       2    1           530624                40.0   40.0   
20        1    1           530157                39.9   39.9   
..      ...  ...              ...                 ...    ...   
274       1    1           534624                 NaN    NaN   
282       1    1           527544                 NaN    NaN   
288       1    1           529428                 NaN    NaN   
293       1    1           534004                 NaN    NaN   
295       1    1           533886                 NaN    NaN   

     temperature of extremities  abdominal distension  total protein  
259                         3.0                   1.0            7.4  
54                          3.0                   3.0            8.1  
91                          3.0                   3.0            8.1  
281                         3.0                   1.0            6.5  
20                          1.0                   4.0            6.1  
..                          ...                   ...            ...  
274                         NaN                   4.0            NaN  
282                         3.0                   2.0            7.5  
288                         NaN                   NaN            NaN  
293                         3.0                   NaN           62.0  
295                         4.0                   NaN           65.0  

[300 rows x 8 columns]

**Выбросы есть и на нижней границе и на верхней показателей ректальной температуры и пульса. Мое сугубо субъективное мнение - с выбросами ничего не делать, так как значения похожи на правду. Поэтому я бы уточнила данные, а на данный момент оставлю.**

**Температура конечностей**

In [33]:
outliers('temperature of extremities')

surgery  age  hospital number  rectal temperature  pulse  \
89        2    1           535381                38.3   38.3   
71        1    1           529424                37.8   37.8   
215       1    1           529135                37.8   37.8   
69        2    1           530276                38.0   38.0   
68        2    1           529777                 NaN    NaN   
..      ...  ...              ...                 ...    ...   
108       1    1           533696                38.7   38.7   
142       1    1           530402                 NaN    NaN   
45        2    1           529827                 NaN    NaN   
151       1    1           529399                39.3   39.3   
240       1    1           529160                38.0   38.0   

     temperature of extremities  abdominal distension  total protein  
89                          1.0                   1.0           61.0  
71                          1.0                   1.0            5.9  
215                         1.0                   1.0            7.0  
69                          1.0                   NaN            6.3  
68                          1.0                   NaN            7.2  
..                          ...                   ...            ...  
108                         4.0                   4.0           64.0  
142                         4.0                   4.0            4.5  
45                          4.0                   4.0            6.5  
151                         4.0                   2.0            NaN  
240                         4.0                   4.0            6.1  

[244 rows x 8 columns]

**Вздутие живота**

In [34]:
outliers('abdominal distension')

surgery  age  hospital number  rectal temperature  pulse  \
104       1    1           527563                37.8   37.8   
143       1    1           529424                37.8   37.8   
225       2    1           528931                38.4   38.4   
256       1    1           533871                37.5   37.5   
224       1    1           535392                38.8   38.8   
..      ...  ...              ...                 ...    ...   
55        1    1          5282839                38.6   38.6   
169       1    1           535085                 NaN    NaN   
230       1    1           534280                38.5   38.5   
193       1    1           529373                 NaN    NaN   
0         2    1           530101                38.5   38.5   

     temperature of extremities  abdominal distension  total protein  
104                         1.0                   1.0            6.6  
143                         1.0                   1.0            5.9  
225                         1.0                   1.0            7.2  
256                         2.0                   1.0           65.0  
224                         1.0                   1.0           58.0  
..                          ...                   ...            ...  
55                          3.0                   4.0            NaN  
169                         NaN                   4.0            NaN  
230                         3.0                   4.0            NaN  
193                         3.0                   4.0            7.6  
0                           3.0                   4.0            8.4  

[244 rows x 8 columns]

**В этих двух показателях выбросов нет.**

**С пропусками все намного сложнее. Их 30 процентов (как мы знаем из описания к датафрейм). Они располагаются рандомно. Убирать все абсолютно нулевые значения не имеет смысла. Заполнять их близлежащими значениями, нецелесообразно. Допустим мы отсортируем данные по ректальной температуре и нам нужно заполнить пропуски в столбце с температурой конечностей. По тем данным, что имеются, я могу сказать, что при одной и той же ректальной температуре температуры конечностей разных особей отличаются. Единственное, что коррелируется это ректальная температура и пульс, значения идентичны, что возникает вопрос - "а не вставили значения температуры в столбец пульса?". Не считаю нужным заполнять модой/средним - я не смогу гарантировать, что статистические показатели верны, к тому же это медицинские данные. Поэтому я думаю, что правильнее игнорировать пропуски (с чем методы пандас успешно справляются).**

In [40]:
cut_data.sort_values('temperature of extremities', ascending=True).head(10)

surgery  age  hospital number  rectal temperature  pulse  \
89        2    1           535381                38.3   38.3   
71        1    1           529424                37.8   37.8   
215       1    1           529135                37.8   37.8   
69        2    1           530276                38.0   38.0   
68        2    1           529777                 NaN    NaN   
67        2    1           528931                38.4   38.4   
66        2    1           534784                37.9   37.9   
255       1    1          5294539                38.8   38.8   
73        1    1           528469                38.1   38.1   
212       1    1          5294369                38.9   38.9   

     temperature of extremities  abdominal distension  total protein  
89                          1.0                   1.0           61.0  
71                          1.0                   1.0            5.9  
215                         1.0                   1.0            7.0  
69                          1.0                   NaN            6.3  
68                          1.0                   NaN            7.2  
67                          1.0                   1.0            7.2  
66                          1.0                   1.0           74.0  
255                         1.0                   3.0            3.3  
73                          1.0                   1.0            7.5  
212                         1.0                   3.0            6.3

**На примере этих 10 лошадей с одинаковой температурой конечностей я могу сказать, что остальные показатели диаметрально разные. Может зависимость и есть, но ее можно выяснить только с помощью коэффициентов корреляции.**